finality-aleph/src/network/substrate.rs

```diff
@@ -60,163 +82,230 @@ impl ProtocolNaming {
-                            }
-                            continue;
-                        }
-                        PeerDisconnected(remote) => {
-                            trace!(target: "aleph-network", "Disconnected event for peer {:?}", remote);
-                            let addresses: Vec<_> = iter::once(remote).collect();
-                            if let Err(e) = self.network.remove_peers_from_reserved_set(
-                                self.naming.protocol_name(&Protocol::Authentication),
-                                addresses.clone(),
-                            ) {
-                                warn!(target: "aleph-network", "Error while removing peer from Protocol::Authentication reserved set: {}", e)
-                            }
-                            if let Err(e) = self.network.remove_peers_from_reserved_set(
-                                self.naming.protocol_name(&Protocol::BlockSync),
-                                addresses,
-                            ) {
-                                warn!(target: "aleph-network", "Error while removing peer from Protocol::BlockSync reserved set: {}", e)
-                            }
-                            continue;
+impl ProtocolNetwork {
+    pub fn new(service: Box<dyn sc_network::config::NotificationService>) -> Self {
+        Self {
+            service,
+            connected_peers: HashSet::new(),
+            last_status_report: time::Instant::now(),
+        }
+    }
+
+    fn random_peer<'a>(&'a self, peer_ids: &'a HashSet<PeerId>) -> Option<&'a PeerId> {
+        peer_ids
+            .intersection(&self.connected_peers)
+            .choose(&mut thread_rng())
+            .or_else(|| self.connected_peers.iter().choose(&mut thread_rng()))
+    }
+
+    fn handle_network_event(&mut self, event: SubstrateEvent) -> Option<(Vec<u8>, PeerId)> {
+        use SubstrateEvent::*;
+        match event {
+            ValidateInboundSubstream {
+                peer: _,
+                handshake: _,
+                result_tx,
+            } => {
+                let _ = result_tx.send(ValidationResult::Accept);
+                None
+            }
+            NotificationStreamOpened { peer, .. } => {
+                self.connected_peers.insert(peer);
+                None
+            }
+            NotificationStreamClosed { peer } => {
+                self.connected_peers.remove(&peer);
+                None
+            }
+            NotificationReceived { peer, notification } => Some((notification, peer)),
+        }
+    }
+
+    fn status_report(&self) {
+        let mut status = String::from("Network status report: ");
+        status.push_str(&format!(
+            "{} connected peers - {:?}; ",
+            self.service.protocol(),
+            self.connected_peers.len()
+        ));
+        info!(target: LOG_TARGET, "{}", status);
+    }
+}
+
+#[derive(Debug)]
+pub enum ProtocolNetworkError {
+    NetworkStreamTerminated,
```

Comment:
Isnt it the same error as before? Cant we use one enum `NetworkError` for example?

Reply:
you mean `SyncNetworkServiceError`?
`SyncNetworkService` will be removed soon + these are 2 different kinds of networks

Reply:
ok, but the error is kinda the same? ANyway, feel free to resolve or go with my suggestion since one of them will be removed anyway